util/db/cluster.go

```diff
@@ -106,71 +122,79 @@ type ClusterEvent struct {
+	namespace string,
+	handleAddEvent func(cluster *appv1.Cluster),
+	handleModEvent func(oldCluster *appv1.Cluster, newCluster *appv1.Cluster),
+	handleDeleteEvent func(clusterServer string)) error {
 	localCls, err := db.GetCluster(ctx, common.KubernetesInternalAPIServerAddr)
 	if err != nil {
 		return err
 	}
+	handleAddEvent(localCls)
 
-	defer w.Stop()
-	watchErr := make(chan error)
-
-	// trigger callback with event for local cluster since it always considered added
-	callback(&ClusterEvent{Type: watch.Added, Cluster: localCls})
-
-	go func() {
-		for next := range w.ResultChan() {
-			secret, ok := next.Object.(*apiv1.Secret)
-			if !ok {
-				watchErr <- fmt.Errorf("failed to convert event object to secret")
-				return
-			}
-			cluster := secretToCluster(secret)
-
-			// change local cluster event to modified or deleted, since it cannot be re-added or deleted
-			if cluster.Server == common.KubernetesInternalAPIServerAddr {
-				if next.Type == watch.Deleted {
-					next.Type = watch.Modified
-					cluster = &localCluster
-				} else if next.Type == watch.Added {
-					if !reflect.DeepEqual(localCls.Config, cluster.Config) {
-						localCls = cluster
-						next.Type = watch.Modified
-					} else {
-						continue
+	clusterSecretListOptions := func(options *metav1.ListOptions) {
+		clusterLabelSelector := fields.ParseSelectorOrDie(common.LabelKeySecretType + "=" + common.LabelValueSecretTypeCluster)
+		options.LabelSelector = clusterLabelSelector.String()
+	}
+	clusterEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if secretObj, ok := obj.(*apiv1.Secret); ok {
+				cluster, err := db.GetCluster(context.Background(), string(secretObj.Data["server"]))
+				if err != nil {
+					return
+				}
+				if cluster.Server == common.KubernetesInternalAPIServerAddr {
+					if reflect.DeepEqual(localCls.Config, cluster.Config) {
+						return
 					}
-				} else {
 					localCls = cluster
+					// change local cluster event to modified or deleted, since it cannot be re-added or deleted
+					handleModEvent(nil, localCls)
+					return
 				}
+				handleAddEvent(cluster)
 			}
-
-			callback(&ClusterEvent{
-				Type:    next.Type,
-				Cluster: cluster,
-			})
-		}
-		watchErr <- errors.New("secret watch closed")
+		},
+		DeleteFunc: func(obj interface{}) {
+			if secretObj, ok := obj.(*apiv1.Secret); ok {
+				if string(secretObj.Data["server"]) == common.KubernetesInternalAPIServerAddr {
+					// change local cluster event to modified or deleted, since it cannot be re-added or deleted
+					handleModEvent(nil, localCls)
+				} else {
+					handleDeleteEvent(string(secretObj.Data["server"]))
+				}
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if oldSecretObj, ok := oldObj.(*apiv1.Secret); ok {
+				if newSecretObj, ok := newObj.(*apiv1.Secret); ok {
+					oldCluster, err := db.GetCluster(context.Background(), string(oldSecretObj.Data["server"]))
```

Comment:
`db.GetCluster` reads from the same in-memory cache so `oldCluster` and `newCluster` variables has the same object. Instead we need to reconstruct cluster object from `newSecretObj` and `oldSecretObj`.